esp-hal/src/soc/mod.rs

```diff
@@ -78,82 +78,117 @@ fn hal_main(a0: usize, a1: usize, a2: usize) -> ! {
+        .literal sym_rtc_slow_persistent_start, {_rtc_slow_persistent_start}
+        ",
+        __init_persistent = sym __init_persistent,
+        _xtensa_lx_rt_zero_fill = sym _xtensa_lx_rt_zero_fill,
+
+        _rtc_fast_bss_end = sym _rtc_fast_bss_end,
+        _rtc_fast_bss_start = sym _rtc_fast_bss_start,
+        _rtc_fast_persistent_end = sym _rtc_fast_persistent_end,
+        _rtc_fast_persistent_start = sym _rtc_fast_persistent_start,
+
+        _rtc_slow_bss_end = sym _rtc_slow_bss_end,
+        _rtc_slow_bss_start = sym _rtc_slow_bss_start,
+        _rtc_slow_persistent_end = sym _rtc_slow_persistent_end,
+        _rtc_slow_persistent_start = sym _rtc_slow_persistent_start,
+    );
+
+    #[unsafe(export_name = "__post_init")]
+    #[unsafe(naked)]
+    #[allow(named_asm_labels)]
+    extern "C" fn post_init() {
+        naked_asm!(
+            "
+            entry a1, 0
+
+            l32r   a6, sym_xtensa_lx_rt_zero_fill      // Pre-load address of zero-fill function
+
+            l32r   a10, sym_rtc_fast_bss_start         // Set input range to .rtc_fast.bss
+            l32r   a11, sym_rtc_fast_bss_end           //
+            callx8 a6                                  // Zero-fill
+
+            l32r   a10, sym_rtc_slow_bss_start         // Set input range to .rtc_slow.bss
+            l32r   a11, sym_rtc_slow_bss_end           //
+            callx8 a6                                  // Zero-fill
+
+            l32r   a5,  sym_init_persistent            // Do we need to initialize persistent data?
+            callx8 a5
+            beqz.n a10, .Lpost_init_return             // If not, skip initialization
+
+            l32r   a10, sym_rtc_fast_persistent_start  // Set input range to .rtc_fast.persistent
+            l32r   a11, sym_rtc_fast_persistent_end    //
+            callx8 a6                                  // Zero-fill
+
+            l32r   a10, sym_rtc_slow_persistent_start  // Set input range to .rtc_slow.persistent
+            l32r   a11, sym_rtc_slow_persistent_end    //
+            callx8 a6                                  // Zero-fill
+
+        .Lpost_init_return:
+            retw.n
+        ",
+        )
     }
 
-    // copying data from flash to various data segments is done by the bootloader
-    // initialization to zero needs to be done by the application
-
-    // Initialize RTC RAM
-    unsafe {
-        xtensa_lx_rt::zero_bss(
-            core::ptr::addr_of_mut!(_rtc_fast_bss_start),
-            core::ptr::addr_of_mut!(_rtc_fast_bss_end),
-        );
-        xtensa_lx_rt::zero_bss(
-            core::ptr::addr_of_mut!(_rtc_slow_bss_start),
-            core::ptr::addr_of_mut!(_rtc_slow_bss_end),
-        );
-    }
-    if matches!(
-        crate::system::reset_reason(),
-        None | Some(crate::rtc_cntl::SocResetReason::ChipPowerOn)
-    ) {
+    #[xtensa_lx_rt::pre_init]
+    #[cfg_attr(esp32s3, ram)]
+    unsafe fn esp32_reset() {
```

Comment:
Technically this and its callees should also be naked functions, but I'm not too bothered by this, as we don't touch any memory here.